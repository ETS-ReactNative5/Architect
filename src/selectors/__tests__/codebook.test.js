/* eslint-env jest */

import testState from '../../__tests__/testState.json';
import {
  asOption,
  propertyMaps,
  getTypes,
  getType,
  getVariables,
  getVariable,
  getVariableOptions,
} from '../codebook';

const getOptions = mergeProps => ({
  includeDraft: false,
  ...mergeProps,
});

describe('codebook selectors', () => {
  describe('getTypes', () => {
    it('returns list of types', () => {
      const types = getTypes(testState);
      expect(types).toHaveLength(6);
      expect(types).toMatchSnapshot();
    });
  });

  describe('getType', () => {
    it('returns type object with meta', () => {
      const subject = { entity: 'node', type: 'bar' };
      const options = getOptions({ subject });
      const type = getType(testState, options);
      expect(type).toHaveProperty('properties');
      expect(type).toHaveProperty('subject', subject);
      expect(type).toMatchSnapshot();
    });
  });

  describe('getVariables', () => {
    it('gets all variables from the codebook with meta data', () => {
      const variables = getVariables(testState);
      expect(variables).toHaveLength(12);
      variables.forEach((variable) => {
        expect(variable).toHaveProperty('id');
        expect(variable).toHaveProperty('properties');
        expect(variable).toHaveProperty('subject');
        expect(variable).toHaveProperty('inUse');
        expect(variable).toHaveProperty('usage');
        expect(variable).toMatchSnapshot();
      });
    });
  });

  describe('getVariable', () => {
    const id = 'alpha';
    const options = getOptions({ id });
    const variable = getVariable(testState, options);
    expect(variable).toHaveProperty('properties');
    expect(variable).toHaveProperty('subject');
    expect(variable).toHaveProperty('inUse');
    expect(variable).toHaveProperty('usage');
    expect(variable).toMatchSnapshot();
  });

  describe('getVariableOptions', () => {
    const id = '1234-1234-2';
    const options = getOptions({ id });
    const variableOptions = getVariableOptions(testState, options);
    expect(variableOptions).toHaveLength(2);
    expect(variableOptions).toMatchSnapshot();
  });

  describe('asOption', () => {
    const types = getTypes(testState);

    const typeOptions = types.map(asOption(propertyMaps.entity));
    typeOptions.forEach((option) => {
      expect(option).toHaveProperty('value');
      expect(option).toHaveProperty('label');
      expect(option).toHaveProperty('color');
    });

    const variables = getVariables(testState);
    const variableOptions = variables.map(asOption());
    variableOptions.forEach((option) => {
      expect(option).toHaveProperty('value');
      expect(option).toHaveProperty('label');
      expect(option).toHaveProperty('type');
    });
  });
});
