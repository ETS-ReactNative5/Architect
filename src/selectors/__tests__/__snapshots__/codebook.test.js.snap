// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
Object {
  "id": "alpha",
  "inUse": true,
  "properties": Object {
    "component": "Text",
    "name": "ALPHA",
    "type": "text",
  },
  "subject": Object {
    "entity": "node",
    "type": "bar",
  },
  "usage": Array [
    "stages[0].form.fields[0].variable",
  ],
}
`;

exports[` 2`] = `
Array [
  Object {
    "label": "One",
    "value": "one",
  },
  Object {
    "label": "Two",
    "value": "two",
  },
]
`;

exports[`codebook selectors getType returns type object with meta 1`] = `
Object {
  "properties": Object {
    "color": "blue",
    "name": "bar",
    "variables": Object {
      "1234-1234-4": Object {
        "name": "1234-1234-4",
        "type": "text",
      },
      "1234-1234-5": Object {
        "name": "1234-1234-5",
        "type": "text",
      },
      "alpha": Object {
        "component": "Text",
        "name": "ALPHA",
        "type": "text",
      },
      "bravo": Object {
        "name": "BRAVO",
        "type": "text",
      },
      "charlie": Object {
        "name": "CHARLIE",
        "type": "location",
      },
    },
  },
  "subject": Object {
    "entity": "node",
    "type": "bar",
  },
}
`;

exports[`codebook selectors getTypes returns list of types 1`] = `
Array [
  Object {
    "properties": Object {
      "variables": Object {
        "sierra": Object {
          "name": "SIERRA",
          "type": "text",
        },
        "tango": Object {
          "name": "TANGO",
          "type": "text",
        },
      },
    },
    "subject": Object {
      "entity": "ego",
    },
  },
  Object {
    "properties": Object {
      "color": "black",
      "name": "foo",
      "variables": Object {},
    },
    "subject": Object {
      "entity": "node",
      "type": "foo",
    },
  },
  Object {
    "properties": Object {
      "color": "black",
      "name": "pop",
      "variables": Object {},
    },
    "subject": Object {
      "entity": "node",
      "type": "pop",
    },
  },
  Object {
    "properties": Object {
      "color": "blue",
      "name": "bar",
      "variables": Object {
        "1234-1234-4": Object {
          "name": "1234-1234-4",
          "type": "text",
        },
        "1234-1234-5": Object {
          "name": "1234-1234-5",
          "type": "text",
        },
        "alpha": Object {
          "component": "Text",
          "name": "ALPHA",
          "type": "text",
        },
        "bravo": Object {
          "name": "BRAVO",
          "type": "text",
        },
        "charlie": Object {
          "name": "CHARLIE",
          "type": "location",
        },
      },
    },
    "subject": Object {
      "entity": "node",
      "type": "bar",
    },
  },
  Object {
    "properties": Object {
      "color": "red",
      "name": "1234-1234-1234",
      "variables": Object {
        "1234-1234-1": Object {
          "name": "my layout",
          "type": "layout",
        },
        "1234-1234-2": Object {
          "name": "my category",
          "options": Array [
            Object {
              "label": "One",
              "value": "one",
            },
            Object {
              "label": "Two",
              "value": "two",
            },
          ],
          "type": "categorical",
        },
        "1234-1234-3": Object {
          "name": "my boolean",
          "type": "boolean",
        },
        "1234-1234-4": Object {
          "name": "other boolean 2",
          "type": "boolean",
        },
        "1234-1234-5": Object {
          "name": "other boolean 4",
          "type": "boolean",
        },
      },
    },
    "subject": Object {
      "entity": "node",
      "type": "1234-1234-1234",
    },
  },
  Object {
    "properties": Object {
      "color": "blue",
      "name": "an edge",
    },
    "subject": Object {
      "entity": "edge",
      "type": "1234-5",
    },
  },
]
`;

exports[`codebook selectors getVariables gets all variables from the codebook with meta data 1`] = `
Object {
  "id": "sierra",
  "inUse": false,
  "properties": Object {
    "name": "SIERRA",
    "type": "text",
  },
  "subject": Object {
    "entity": "ego",
  },
  "usage": Array [],
}
`;

exports[`codebook selectors getVariables gets all variables from the codebook with meta data 2`] = `
Object {
  "id": "tango",
  "inUse": false,
  "properties": Object {
    "name": "TANGO",
    "type": "text",
  },
  "subject": Object {
    "entity": "ego",
  },
  "usage": Array [],
}
`;

exports[`codebook selectors getVariables gets all variables from the codebook with meta data 3`] = `
Object {
  "id": "alpha",
  "inUse": true,
  "properties": Object {
    "component": "Text",
    "name": "ALPHA",
    "type": "text",
  },
  "subject": Object {
    "entity": "node",
    "type": "bar",
  },
  "usage": Array [
    "stages[0].form.fields[0].variable",
  ],
}
`;

exports[`codebook selectors getVariables gets all variables from the codebook with meta data 4`] = `
Object {
  "id": "bravo",
  "inUse": true,
  "properties": Object {
    "name": "BRAVO",
    "type": "text",
  },
  "subject": Object {
    "entity": "node",
    "type": "bar",
  },
  "usage": Array [
    "stages[1].form.fields[0].variable",
  ],
}
`;

exports[`codebook selectors getVariables gets all variables from the codebook with meta data 5`] = `
Object {
  "id": "charlie",
  "inUse": false,
  "properties": Object {
    "name": "CHARLIE",
    "type": "location",
  },
  "subject": Object {
    "entity": "node",
    "type": "bar",
  },
  "usage": Array [],
}
`;

exports[`codebook selectors getVariables gets all variables from the codebook with meta data 6`] = `
Object {
  "id": "1234-1234-4",
  "inUse": true,
  "properties": Object {
    "name": "1234-1234-4",
    "type": "text",
  },
  "subject": Object {
    "entity": "node",
    "type": "bar",
  },
  "usage": Array [
    "stages[0].form.fields[1].variable",
  ],
}
`;

exports[`codebook selectors getVariables gets all variables from the codebook with meta data 7`] = `
Object {
  "id": "1234-1234-5",
  "inUse": true,
  "properties": Object {
    "name": "1234-1234-5",
    "type": "text",
  },
  "subject": Object {
    "entity": "node",
    "type": "bar",
  },
  "usage": Array [
    "stages[1].form.fields[1].variable",
  ],
}
`;

exports[`codebook selectors getVariables gets all variables from the codebook with meta data 8`] = `
Object {
  "id": "1234-1234-1",
  "inUse": false,
  "properties": Object {
    "name": "my layout",
    "type": "layout",
  },
  "subject": Object {
    "entity": "node",
    "type": "1234-1234-1234",
  },
  "usage": Array [],
}
`;

exports[`codebook selectors getVariables gets all variables from the codebook with meta data 9`] = `
Object {
  "id": "1234-1234-2",
  "inUse": false,
  "properties": Object {
    "name": "my category",
    "options": Array [
      Object {
        "label": "One",
        "value": "one",
      },
      Object {
        "label": "Two",
        "value": "two",
      },
    ],
    "type": "categorical",
  },
  "subject": Object {
    "entity": "node",
    "type": "1234-1234-1234",
  },
  "usage": Array [],
}
`;

exports[`codebook selectors getVariables gets all variables from the codebook with meta data 10`] = `
Object {
  "id": "1234-1234-3",
  "inUse": false,
  "properties": Object {
    "name": "my boolean",
    "type": "boolean",
  },
  "subject": Object {
    "entity": "node",
    "type": "1234-1234-1234",
  },
  "usage": Array [],
}
`;

exports[`codebook selectors getVariables gets all variables from the codebook with meta data 11`] = `
Object {
  "id": "1234-1234-4",
  "inUse": true,
  "properties": Object {
    "name": "other boolean 2",
    "type": "boolean",
  },
  "subject": Object {
    "entity": "node",
    "type": "1234-1234-1234",
  },
  "usage": Array [
    "stages[0].form.fields[1].variable",
  ],
}
`;

exports[`codebook selectors getVariables gets all variables from the codebook with meta data 12`] = `
Object {
  "id": "1234-1234-5",
  "inUse": true,
  "properties": Object {
    "name": "other boolean 4",
    "type": "boolean",
  },
  "subject": Object {
    "entity": "node",
    "type": "1234-1234-1234",
  },
  "usage": Array [
    "stages[1].form.fields[1].variable",
  ],
}
`;
